Prevent calls to exit implementations in tests

fabianlupa opened the pull request:
From the discussion in #6646. This generally disables calling productive exit implementations in unit tests. Ideally these could also be prevented by being mocked in all tests but having this fallback seems like a good idea to me.

fabianlupa commented:
Of course a discussion on "There now is a check in production code whether it's run in a test or not" would also be very valid ;)

christianguenter2 commented on src/utils/zcl_abapgit_environment.clas.abap:
```diff
@@ -238,4 +238,16 @@ CLASS zcl_abapgit_environment IMPLEMENTATION.
+    " test context which should be fine as there are no unit tests delivered in it.
+    cl_abap_typedescr=>describe_by_name(
+      EXPORTING
+        p_name         = |\\PROGRAM={ sy-repid }\\CLASS=LTCL_TEST|
```
How does that work? Isn't `SY-REPID` always `ZCL_ABAPGIT_ENVIRONMENT=======CP`? What if the class isn't called `LTCL_TEST`?

fabianlupa replied:
`LTCL_TEST` is a test class with the `FOR TESTING ADDITION`. RTTI cannot access test classes unless it is called in a running unit test ;)

christianguenter2 replied:
I had a different understanding of how it should work. Got it now. 
It didn't work for me because I also mocked `ZIF_ABAPGIT_ENVIRONMENT~IS_RUNNING_IN_TEST_CONTEXT` with an empty implementation returning FALSE. Not sure whether that is an issue or not.

fabianlupa replied:
Ah, interesting. Might as well move it out of the interface and directly call the method in the exit factory method or even move it to the exit class. I cannot think of another place where this check would make sense so it doesn't need to be that reusable.

mbtools replied:
+1 for private method

mbtools commented on src/utils/zcl_abapgit_environment.clas.testclasses.abap:
```diff
@@ -63,4 +64,8 @@ CLASS ltcl_test IMPLEMENTATION.
   ENDMETHOD.
 
+  METHOD is_running_in_test_context.
+    cl_abap_unit_assert=>assert_true( mi_cut->is_running_in_test_context( ) ).
```
`assert_equal` required for 702

mbtools commented on src/utils/zcl_abapgit_environment.clas.testclasses.abap:
add comment "if renaming this test class, then is_running_in_test_context needs to be adjusted accordingly"

fabianlupa replied:
Sure. The test method should also conveniently fail in that case.

mbtools reviewed (changes requested):
Technically, the scenario asks for a test environment with various exit implementations. But since we (contributors) use exits in our dev environments, this looks like a valid compromise.

fabianlupa commented:
Moved, downported and added comments. Any hint on what to do regarding the failing transpiled test? Mock the exit in zcl_abapgit_data_supporter? Or is some configuration required?

mbtools reviewed (approved):
open-abap doesn't handle this rtti but needs to run the tests. The "if" does the trick

fabianlupa commented:
Huh, thanks.
I wonder why ABC was chosen though which could be an actual system (?) instead of something like !OA.

larshp commented:
> Huh, thanks. I wonder why ABC was chosen though which could be an actual system (?) instead of something like !OA.

I didnt think about that, suggest opening an issue in https://github.com/abaplint/transpiler
